Add support for HUP, USR1, USR2 signals

hassila opened the pull request:
Motivation:
Some servers want to support customised handling of signals except for simply shutting down.

Added support for USR1/USR2 signal traps for (#64) - extended trap() signature with one more optional argument to allow for signal handler to be run multiple times with usage like this:

```
let alrmDispatchSourceSignal = ServiceLifecycle.trap(signal: ServiceLifecycle.Signal.USR1, handler: { signal in
	    print("signal \(signal)")
	}, cancelAfterTrap: false)```

Existing usage of .trap() is unaffected.

Possibly additional signals (HUP?) should be added too.

Changes:
Added USR1/USR2 to Signal struct to allow for trapping them.
Changed .trap() signature to allow for signal handlers to be run multiple times without reregistration.

ktoso commented:
@swift-server-bot test this please

hassila commented:
I added HUP too.

tomerd commented:
thanks @hassila the `soundness` job reports some formatting issues, see: https://github.com/swift-server/swift-service-lifecycle/blob/main/CONTRIBUTING.md#run-scriptssoundnesssh

tomerd reviewed (changes requested):
one comment re. default and formatting issues

hassila commented on Sources/Lifecycle/Lifecycle.swift:
```diff
@@ -177,11 +177,13 @@ extension ServiceLifecycle {
     ///    - handler: closure to invoke when the signal is captured.
     /// - returns: a `DispatchSourceSignal` for the given trap. The source must be cancelled by the caller.
-    public static func trap(signal sig: Signal, handler: @escaping (Signal) -> Void, on queue: DispatchQueue = .global()) -> DispatchSourceSignal {
+    public static func trap(signal sig: Signal, handler: @escaping (Signal) -> Void, on queue: DispatchQueue = .global(), cancelAfterTrap: Bool = true) -> DispatchSourceSignal {
```
Definitely agree, +1. Only reason was reflex to not affect possible other users of API. Will fix.

hassila replied:
I changed default in 8d379dc - missed mentioning it in commit message.

hassila commented:
@tomerd I tried to run soundness and failed:
jocke@solid /V/j/l/swift-service-lifecycle (hassila-usr-signals)> scripts/soundness.sh 
=> Checking for unacceptable language... okay.
=> Checking linux tests... okay.
=> Checking format... ⏎                                                                                                                                                       jocke@solid /V/j/l/swift-service-lifecycle (hassila-usr-signals) [127]> 

Assuming the return keys references CRLF issues (just a guess) but nothing obvious (file just edit on macOS and Linux with emacs/xcode). I have removed the tabs that were there previously though. 

Committed without tabs and added some documentation.

hassila commented:
Ok, I have tried to clean whitespace as much as possible, but I get this weird output from soundness.sh which is difficult to guess the purpose of:


```
jocke@swift511 ~/l/swift-service-lifecycle (hassila-usr-signals)> scripts/soundness.sh
=> Checking for unacceptable language... okay.
=> Checking linux tests... okay.
=> Checking format... ^J                 
jocke@swift511 ~/l/swift-service-lifecycle (hassila-usr-signals) [127]> 

```

tomerd commented:
@swift-server-bot test this please

tomerd commented:
> Ok, I have tried to clean whitespace as much as possible, but I get this weird output from soundness.sh which is difficult to guess the purpose of:

you can run `swiftformat .` to have it format the project for you

hassila commented:
I tried swift format, but it did changes like:
```
-    public static func eventLoopFuture(_ future: @escaping () -> EventLoopFuture<Void>) -> LifecycleHandler {
+    static func eventLoopFuture(_ future: @escaping () -> EventLoopFuture<Void>) -> LifecycleHandler {

```

will look at it  bit more, but seemed a bit aggressive there :-)

tomerd commented:
make sure you use `swiftformat` not `swift-format`: `brew install swiftformat `

hassila commented:
I downloaded and built this one on Linux: https://github.com/nicklockwood/SwiftFormat - not the right one?

hassila commented:
jocke@swift511 ~/l/swift-service-lifecycle (hassila-usr-signals) [1]> swiftformat --version
0.47.12

tomerd commented:
tool looks fine - I think ci uses 0.44.6

you can see the formatting issue in the log: https://ci.swiftserver.group/job/swift-service-lifecycle-soundness-prb/189/console

hassila commented:
Thanks, I couldn't find that log. Pushed fixes for those now. Seems the new version of swift formatter does a bit more, I will try to fetch 0.44.6 to try to get same output (until CI is updated).

hassila commented:
soundness.sh now passed for formatting at least, thanks for the patience answering questions.

tomerd commented:
@swift-server-bot test this please

tomerd commented:
almost there: https://ci.swiftserver.group/job/swift-service-lifecycle-soundness-prb/197/consoleText

hassila commented:
Fixed in 265fb15.

tomerd commented:
@swift-server-bot test this please

hassila commented:
(and installed exact version of swiftformat as used by CI for the future - 0.44.6 doesn't touch as aggressively and can and will now be used... thanks.)

tomerd commented:
thanks @hassila
